pinot-controller/src/main/java/org/apache/pinot/controller/helix/RealtimeConsumerMonitor.java

```diff
@@ -0,0 +1,122 @@
+import org.apache.pinot.controller.util.ConsumingSegmentInfoReader;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.stream.PartitionLagState;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class RealtimeConsumerMonitor extends ControllerPeriodicTask<RealtimeConsumerMonitor.Context> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RealtimeConsumerMonitor.class);
+  private final ConsumingSegmentInfoReader _consumingSegmentInfoReader;
+
+  @VisibleForTesting
+  public RealtimeConsumerMonitor(ControllerConf controllerConf, PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, ControllerMetrics controllerMetrics,
+      ConsumingSegmentInfoReader consumingSegmentInfoReader) {
+    super("RealtimeConsumerMonitor", controllerConf.getRealtimeConsumerMonitorRunFrequency(),
+        controllerConf.getRealtimeConsumerMonitorInitialDelayInSeconds(), pinotHelixResourceManager,
+        leadControllerManager, controllerMetrics);
+    _consumingSegmentInfoReader = consumingSegmentInfoReader;
+  }
+
+  public RealtimeConsumerMonitor(ControllerConf controllerConf, PinotHelixResourceManager pinotHelixResourceManager,
+      LeadControllerManager leadControllerManager, ControllerMetrics controllerMetrics,
+      ExecutorService executorService) {
+    this(controllerConf, pinotHelixResourceManager, leadControllerManager, controllerMetrics,
+        new ConsumingSegmentInfoReader(executorService, new SimpleHttpConnectionManager(), pinotHelixResourceManager));
+  }
+
+  @Override
+  protected void setUpTask() {
+    LOGGER.info("Setting up RealtimeConsumerMonitor task");
+  }
+
+  @Override
+  protected void processTable(String tableNameWithType) {
+    if (!TableType.REALTIME.equals(TableNameBuilder.getTableTypeFromTableName(tableNameWithType))) {
+      return;
+    }
+    try {
+      ConsumingSegmentInfoReader.ConsumingSegmentsInfoMap segmentsInfoMap =
+          _consumingSegmentInfoReader.getConsumingSegmentsInfo(tableNameWithType, 10000);
+      Map<String, List<Long>> partitionToLagSet = new HashMap<>();
+      Map<String, List<Long>> partitionToAvailabilityLagSet = new HashMap<>();
+
+      for (List<ConsumingSegmentInfoReader.ConsumingSegmentInfo> info
+          : segmentsInfoMap._segmentToConsumingInfoMap.values()) {
+        info.forEach(segment -> {
+          segment._partitionOffsetInfo._recordsLagMap.forEach((k, v) -> {
+            if (!PartitionLagState.NOT_CALCULATED.equals(v)) {
+              try {
+                long recordsLag = Long.parseLong(v);
+                partitionToLagSet.computeIfAbsent(k, k1 -> new ArrayList<>());
+                partitionToLagSet.get(k).add(recordsLag);
+              } catch (NumberFormatException nfe) {
+                // skip this as we are unable to parse the lag string
```

Comment:
What can be the scenario of this? I guess that we don't want to log here due to too many lines?

Reply:
There are connectors that don't yet calculate lag or may not have a lag value (for example, kinesis will only have availability lag and not records lag). While the default for any lag value is `NOT_CALCULATED`, it could be anything that the connector plugin returns. Just safe-guarding against invalid long values here. 

I can add a log line here to indicate this exception happened. printing stacktrace will create too many lines.

Reply:
I didn't add a log line here as it is likely to get very noisy. We can always cross-check by running the same query thought our rest api `/consumingSegmentsApi`

Reply:
thanks!